torch/distributed/tensor/_dispatch.py

```diff
@@ -307,8 +306,11 @@ def redistribute_local_args(
                 else:
                     new_local_args.append(local_tensor)
             else:
-                new_local_args.append(reshard_arg_spec)
-
+                # Due to the strategy hashing, we should use the
+                # [args/kwargs]_schema that is not DTensorSpec from
+                # op_info.schema instead of
+                # info.output_sharding.redistribute_schema.
+                new_local_args.append(arg_spec)
```

Comment:
iiuc, your claim is that `arg_spec` comes from the current runtime call, while `reshard_arg_spec` comes from the cache from the first time someone called DTensor with the same cache-key?

let me think about this for 2 cases

1) `arg_spec` was not part of the cache key for this op: then using the value copied from the cache seems wrong and this seems to fix it
2) `arg_spec` was part of the cache key: then it should have worked with the old code, but the new code is fine too

does that sound right?  well, if so, i'm surprised we didn't get bit by this bug before!